src/lib.rs

```diff
@@ -0,0 +1,360 @@
+        Ok(attestation_result)
+    }
+
+    fn new_session(&self, nonce: &Nonce) -> Result<(String, ChallengeResponseSession), Error> {
+        // ask veraison for a new session object
+        let resp = self
+            .new_session_request(nonce)
+            .map_err(|e| Error::ApiError(e.to_string()))?;
+
+        // expect 201 and a Location header containing the URI of the newly
+        // allocated session
+        match resp.status() {
+            reqwest::StatusCode::CREATED => (),
+            status => {
+                // on error the body is a RFC7807 problem detail
+                //
+                // NOTE(tho) -- this assumption does not hold in general because
+                // the request may be intercepted (and dealt with) by HTTP
+                // middleware that is unaware of the API.  We need something
+                // more robust here that dispatches based on the Content-Type
+                // header.
+                let pd: ProblemDetails = resp.json().map_err(|e| Error::ApiError(e.to_string()))?;
+
+                return Err(Error::ApiError(format!(
+                    "newSession response has unexpected status: {}.  Details: {}",
+                    status, pd.detail
+                )));
+            }
+        };
+
+        // extract location header
+        let loc = resp
+            .headers()
+            .get("location")
+            .ok_or(Error::ApiError(
+                "cannot determine URI of the session resource".to_string(),
+            ))?
+            .to_str()
+            .map_err(|e| Error::ApiError(e.to_string()))?;
+
+        // join relative location with base URI
+        let session_url = resp
+            .url()
+            .join(loc)
+            .map_err(|e| Error::ApiError(e.to_string()))?;
+
+        // decode returned session object
+        let crs: ChallengeResponseSession =
+            resp.json().map_err(|e| Error::ApiError(e.to_string()))?;
+
+        Ok((session_url.to_string(), crs))
+    }
+
+    fn new_session_request(&self, nonce: &Nonce) -> Result<reqwest::blocking::Response, Error> {
+        let c = &self.http_client;
+        let u = self.new_session_request_url(nonce)?;
+
+        return c
+            .post(u.as_str())
+            .header(reqwest::header::ACCEPT, CRS_MEDIA_TYPE)
+            .send()
+            .map_err(|e| Error::ApiError(e.to_string()));
+    }
+
+    fn new_session_request_url(&self, nonce: &Nonce) -> Result<url::Url, Error> {
+        let base = &self.base_url;
+
+        let mut new_session_url = base
+            .join("newSession")
+            .map_err(|e| Error::ConfigError(e.to_string()))?;
+
+        let mut q_params = String::new();
+
+        match nonce {
+            Nonce::Value(val) => {
+                if val.len() > 0 {
+                    q_params.push_str("nonce=");
+                    q_params.push_str(&base64::encode_config(val, base64::URL_SAFE));
+                }
+            }
```

Comment:
```suggestion
            Nonce::Value(val) if val.len() > 0 => {
                q_params.push_str("nonce=");
                q_params.push_str(&base64::encode_config(val, base64::URL_SAFE));
            }
```

... however, what happens if `val` is empty, or if `val` is 0 below? No nonce used?

Also, if you use the pattern above, suddenly you won't be covering all cases in the `match`, so you'll need to add a no-op default branch at the end of the `match`, `_ => {}` (if you don't want to do anything when the nonce is empty).

Reply:
> ... however, what happens if `val` is empty, or if `val` is 0 below? No nonce used?

in that case, since no nonce information has been provided, no nonce-related query parameter is built, and veraison applies the defaults.

> Also, if you use the pattern above, suddenly you won't be covering all cases in the `match`, so you'll need to add a no-op default branch at the end of the `match`, `_ => {}` (if you don't want to do anything when the nonce is empty).

The thing I don't understand is how can Nonce be empty?  ISTM that the compiler shouldn't let you use uninitialised variable, right?

Reply:
> The thing I don't understand is how can Nonce be empty? ISTM that the compiler shouldn't let you use an uninitialised variable, right?

You can still do something like `Nonce::Value(Vec::new())` - at least that's what I meant by "empty" above, not that you don't have a `Nonce` variant.

Reply:
> > The thing I don't understand is how can Nonce be empty? ISTM that the compiler shouldn't let you use an uninitialised variable, right?
> 
> You can still do something like `Nonce::Value(Vec::new())`

sure, but that is handled via the `val.len() > 0` test.

Reply:
BTW, your suggestion upthread is captured in [be6ebbb](https://github.com/veraison/rust-apiclient/pull/4/commits/be6ebbb506f46bd02b947493629f4852ec36dae0)

Reply:
> sure, but that is handled via the `val.len() > 0` test.

Yes - my worry was about the "fallback", default arm of the `match`, but that's covered now 😄